service/src/main/kotlin/fi/espoo/evaka/invoicing/service/InvoiceGenerator.kt

```diff
@@ -63,23 +80,47 @@ class InvoiceGenerator(private val draftInvoiceGenerator: DraftInvoiceGenerator)
 
         val codebtors = unhandledDecisions.mapValues { (_, decisions) -> getInvoiceCodebtor(tx, decisions) }
 
-        val invoices =
-            draftInvoiceGenerator.generateDraftInvoices(
-                unhandledDecisions,
-                unhandledPlacements,
-                range,
-                daycareCodes,
-                operationalDays,
-                feeThresholds,
-                absences,
-                freeChildren,
-                codebtors
-            )
+        return InvoiceCalculationData(
+            decisions = unhandledDecisions,
+            placements = unhandledPlacements,
+            period = range,
+            daycareCodes = daycareCodes,
+            operationalDays = operationalDays,
+            feeThresholds = feeThresholds,
+            absences = absences,
+            freeChildren = freeChildren,
+            codebtors = codebtors
+        )
+    }
 
-        tx.deleteDraftInvoicesByDateRange(range)
-        tx.upsertInvoices(invoices)
+    fun createAllDraftInvoices(tx: Database.Transaction, range: DateRange): List<Invoice> {
+        val invoiceData = calculateInvoiceData(tx, range)
+
+        return draftInvoiceGenerator.generateDraftInvoices(
+            invoiceData.decisions,
+            invoiceData.placements,
+            invoiceData.period,
+            invoiceData.daycareCodes,
+            invoiceData.operationalDays,
+            invoiceData.feeThresholds,
+            invoiceData.absences,
+            invoiceData.freeChildren,
+            invoiceData.codebtors
+        )
     }
 
+    data class InvoiceCalculationData(
+        val decisions: Map<PersonId, List<FeeDecision>>,
+        val placements: Map<PersonId, List<Placements>>,
+        val period: DateRange,
```

Comment:
Nitpicking, but `period` could be removed. It's passed straight into `InvoiceCalculationData` from the function parameter anyway.